// @silver0542/directus-extension-cma-components-editor/src/custom-plugins/plugin-formColTitle.js
export default class FormColTitle {
	// constructor({ data, config, api, block }) {
	constructor(params) {
	

		this.api = params.api;

		this.nodes = {
            wrapper: null,
			wrapperParams: null,
			wrapperClass: null,
			titleClass: null,
			title: null,			
		};

		this.data = params.data;

		// console.log(params);
		

		
	}

	

	static get toolbox() {
		return {
			title: 'Form Col Title',
			icon: `<svg class="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 15v2m-6 4h12a2 2 0 002-2v-6a2 2 0 00-2-2H6a2 2 0 00-2 2v6a2 2 0 002 2zm10-10V7a4 4 0 00-8 0v4h8z"></path></svg>`,
		};
	}

	render() {
 

        this.nodes.wrapper = this.make('div', 'HeadLineForm');

        this.nodes.wrapperClass = this.make('div', 'HeadLineWrapper');

        this.nodes.wrapperParams = this.make('div', 'HeadLineParams');
        
        const paramsWrapperClassWrp = this.make('div', 'HeadLineParams_inputWrapper');
        const paramsWrapperClassLabel = this.make('label',
        '',
        {
            contentEditable: false,
            for: 'wrapperClass',
        },
        'Wrapper Class');
        const paramsWrapperClassInput = this.make('input', '', {
			type: 'text',
			name: 'wrapperClass',
			value: this.data && this.data.wrapperClass ? this.data.wrapperClass : '',
		});


        const paramsTitleClassWrp = this.make('div', 'HeadLineParams_inputWrapper');
        const paramsTitleClassLabel = this.make('label',
        '',
        {
            contentEditable: false,
            for: 'titleClass',
        },
        'Title Class');
        const paramsTitleClassInput = this.make('input', '', {
			type: 'text',
			name: 'titleClass',
			value: this.data && this.data.titleClass ? this.data.titleClass : '',
		});
        
      


        // HEAD LINE
        
        this.nodes.title = this.make(
			'div',
			'HeadLineTitle_inputWrapper',
			{
				contentEditable: true,
			},
			this.data && this.data.title ? this.data.title : 'Step 1 - Select Move Type'
		); 
      



        paramsWrapperClassWrp.appendChild(paramsWrapperClassLabel);
        paramsWrapperClassWrp.appendChild(paramsWrapperClassInput);
        
        paramsTitleClassWrp.appendChild(paramsTitleClassLabel);
        paramsTitleClassWrp.appendChild(paramsTitleClassInput);

        this.nodes.wrapperParams.appendChild(paramsWrapperClassWrp);
        this.nodes.wrapperParams.appendChild(paramsTitleClassWrp);

		this.nodes.wrapperClass.appendChild(this.nodes.title);

		this.nodes.wrapper.appendChild(this.nodes.wrapperClass);
		this.nodes.wrapper.appendChild(this.nodes.wrapperParams);



		return this.nodes.wrapper;
	}

	

	save(toolsContent) {
	
		const wrapperClass = toolsContent.querySelector('[name="wrapperClass"]').value;
		const titleClass = toolsContent.querySelector('[name="titleClass"]').value;
	
        const title = toolsContent.querySelector('.HeadLineTitle_inputWrapper').textContent;
        // const subtitle = toolsContent.querySelector('.HeadLine_inputWrapper').textContent;
        

		/**
		 * Fill missing fields with empty strings
		 */

		// console.log('sAVE::>>>> ', phone, secureIcon);
		Object.assign(this.data, {
		    wrapperClass: wrapperClass.trim() || '',
            titleClass: titleClass.trim() || '',
            title: title.trim() || '',
		}); 
		return this.data;
	}

	


	
	get CSS() {
		return {
			baseClass: this.api.styles.block,
			input: this.api.styles.input,
			loader: this.api.styles.loader,

			/**
			 * Tool's classes
			 */
			wrapperClass: 'cdx-component-HeadLine-WrapperClass',
			wrapperParams: 'cdx-headerParams',
			wrapperHeader: 'cdx-header',
			headerClass: 'cdx-headerParams__headerClass',
			headerWrapperClass: 'cdx-headerParams__WrapperClass',
			showPhone: 'cdx-headerParams__showPhone',
			phoneAsImage: 'cdx-headerParams__phoneAsImage',

			phone: 'cdx-header__phone',
			secureIcon: 'cdx-header__secureIcon',
			name: 'cdx-header__name', 
		};
	}

	validate(savedData) {
		/**
		 * Return false if fields are empty
		 */
		return (
			savedData.wrapperClass ||
			savedData.titleClass ||
			savedData.title 
		);
	}

	make(tagName, classNames = null, attributes = {}, content = '') {
		const el = document.createElement(tagName);

		if (Array.isArray(classNames)) {
			el.classList.add(...classNames);
		} else if (classNames) {
			el.classList.add(classNames);
		}

		for (const attrName in attributes) {
			el[attrName] = attributes[attrName];
		}
		if (content !== '') {
			el.innerHTML = content;
		}

		return el;
	}

	renderSettings() {
		const settings = [
			
			{
				name: 'Custom Configuration',
				title: 'Custom Configuration',
				type: 'showConfig',
				prop: { showSettings: true},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="white"/>
				<path d="M10 6.25V5M10 6.25C9.66848 6.25 9.35054 6.3817 9.11612 6.61612C8.8817 6.85054 8.75 7.16848 8.75 7.5C8.75 7.83152 8.8817 8.14946 9.11612 8.38388C9.35054 8.6183 9.66848 8.75 10 8.75M10 6.25C10.3315 6.25 10.6495 6.3817 10.8839 6.61612C11.1183 6.85054 11.25 7.16848 11.25 7.5C11.25 7.83152 11.1183 8.14946 10.8839 8.38388C10.6495 8.6183 10.3315 8.75 10 8.75M6.25 13.75C6.58152 13.75 6.89946 13.6183 7.13388 13.3839C7.3683 13.1495 7.5 12.8315 7.5 12.5C7.5 12.1685 7.3683 11.8505 7.13388 11.6161C6.89946 11.3817 6.58152 11.25 6.25 11.25M6.25 13.75C5.91848 13.75 5.60054 13.6183 5.36612 13.3839C5.1317 13.1495 5 12.8315 5 12.5C5 12.1685 5.1317 11.8505 5.36612 11.6161C5.60054 11.3817 5.91848 11.25 6.25 11.25M6.25 13.75V15M6.25 11.25V5M10 8.75V15M13.75 13.75C14.0815 13.75 14.3995 13.6183 14.6339 13.3839C14.8683 13.1495 15 12.8315 15 12.5C15 12.1685 14.8683 11.8505 14.6339 11.6161C14.3995 11.3817 14.0815 11.25 13.75 11.25M13.75 13.75C13.4185 13.75 13.1005 13.6183 12.8661 13.3839C12.6317 13.1495 12.5 12.8315 12.5 12.5C12.5 12.1685 12.6317 11.8505 12.8661 11.6161C13.1005 11.3817 13.4185 11.25 13.75 11.25M13.75 13.75V15M13.75 11.25V5" stroke="black" stroke-linecap="round" stroke-linejoin="round"/>
				</svg>												
				`,
			},
        ]

		const wrapper = document.createElement('div');
	    const wrapperAdditional = document.createElement('div');
		
		
		
		// const wrapper_header = document.createElement('div');
		// wrapper_header.classList.add('cdx-settings-button-header-wrapper');
	
		settings.forEach((tune) => {
			let button = document.createElement('div');
			
			button.classList.add('cdx-settings-button');
			button.setAttribute('title', tune.title);
			button.innerHTML = tune.icon;
			wrapperAdditional.appendChild(button);
		
		

			button.addEventListener('click', () => {
				this._toggleTune(tune.type, tune.prop);
				if (tune.type == 'showConfig'){
					for (var item of document.querySelectorAll('.cdx-settings-button--active-FormColTitleContainerSettings')) {
						item.classList.remove('cdx-settings-button--active-FormColTitleContainerSettings');
					}
					button.classList.toggle('cdx-settings-button--active-FormColTitleContainerSettings');
				}			 
			  });
		});
	
		wrapper.appendChild(wrapperAdditional);

		return wrapper;
	}

	_toggleTune(tune, props) {
        if (tune == 'showConfig'){
            this.nodes.wrapper.querySelector('.HeadLineParams').classList.toggle('active')
        }
    }
}
