// @silver0542/directus-extension-cma-components-editor/src/custom-plugins/tune-padding.js


class PaddingTune {

    /**
     * Default alignment
     *
     * @public
     * @returns {string}
     */
    static get DEFAULT_ALIGNMENT() {
        return {};
    }

    static get isTune() {
        return true;
    }

    getAlignment(){
        if(!!this.settings?.blocks && this.settings.blocks.hasOwnProperty(this.block.name)){
            return this.settings.blocks[this.block.name]
        }
        if(!!this.settings?.default){
            return this.settings.default
        }
        return PaddingTune.DEFAULT_ALIGNMENT
    }
    /**
     *
     * @param api
     * @param data 既に設定されているデータ
     * @param settings tuneに設定項目
     * @param block tuneに設定されてるblock
     */
    constructor({ api, data, config, block}) {
        this.api = api;
        this.block = block;
        /**
         config:{
            default: "right",
            blocks: {
              header: 'center',
              list: 'right'
            }
          },
         */
        this.settings = config;
        let defaultVal = {  
                padding: '',
                paddingY: '',
                paddingT: '',
                paddingB: '',
                paddingX: '',
                paddingL: '',
                paddingR: '',
                paddingClass: '',            
        };
        console.log(data);
        this.data = data || { Padding: defaultVal }
       

    }

    /**
     * block自体をwrapしてくれる
     * constructorで与えられたalignmentを代入しようとすると、holderが確定してなく
     * renderでやろうとすると、tuneを表示時に処理が走る
     * @param blockContent
     */
    wrap(blockContent) {
        this.wrapper = this.make("div");
        // this.wrapper.classList.toggle(`${this.data.Padding}`)
        this.wrapper.append(blockContent)
        return this.wrapper
    }

    /**
     * rendering block tune
     * @returns {*}
     */
    render() {
       //  console.log('Data Padding:', this.data);
        const wrapper = this.make("div");

		wrapper.classList.add('cdx-settings-button-header-wrapper');

        var wrap = document.createElement('div');
        wrap.classList.add('MPwrap');
        wrap.classList.add('paddingWrapper');
        wrap.innerHTML = '<p>Padding</p>';

        var wrapP = document.createElement('div');
        wrapP.classList.add('MPwrapPTop');

        var wrapPInput = document.createElement('input');
        wrapPInput.classList.add('MPwrapPTopInput');
        wrapPInput.setAttribute('name', 'padding_top');
        wrapPInput.setAttribute('type', 'number');
        // wrapPInput.setAttribute('min', '0');
        wrapPInput.setAttribute('max', '96');
        wrapPInput.setAttribute('step', '0.5');
        wrapPInput.setAttribute('value', this.data.Padding.paddingT ? this.data.Padding.paddingT : '');

        wrapP.appendChild(wrapPInput);
        wrap.appendChild(wrapP);


        var wrapP2 = document.createElement('div');
        wrapP2.classList.add('MPwrapPSide');

        wrapPInput = document.createElement('input');
        wrapPInput.classList.add('MPwrapPTopInput_LEFT');
        wrapPInput.setAttribute('name', 'padding_left');
        wrapPInput.setAttribute('type', 'number');
        // wrapPInput.setAttribute('min', '0');
        wrapPInput.setAttribute('max', '96');
        wrapPInput.setAttribute('step', '0.5');
        wrapPInput.setAttribute('value', this.data.Padding.paddingL ? this.data.Padding.paddingL : '');
            
            
       

        wrapP2.appendChild(wrapPInput);

        wrapPInput = document.createElement('input');
        wrapPInput.classList.add('MPwrapPTopInput');
        wrapPInput.setAttribute('name', 'padding');
        wrapPInput.setAttribute('type', 'text');
        // wrapPInput.setAttribute('min', '0');
        wrapPInput.setAttribute('max', '96');
        wrapPInput.setAttribute('step', '0.5');
        wrapPInput.setAttribute('value', this.data.Padding.padding ? this.data.Padding.padding : '');

        wrapP2.appendChild(wrapPInput);

        wrapPInput = document.createElement('input');
        wrapPInput.classList.add('MPwrapPTopInput_RIGHT');
        wrapPInput.setAttribute('name', 'padding_right');
        wrapPInput.setAttribute('type', 'number');
        // wrapPInput.setAttribute('min', '0');
        wrapPInput.setAttribute('max', '96');
        wrapPInput.setAttribute('step', '0.5');
        wrapPInput.setAttribute('value', this.data.Padding.paddingR ? this.data.Padding.paddingR : '');
        wrapP2.appendChild(wrapPInput);


        var wrapP3 = document.createElement('div');
        wrapP3.classList.add('MPwrapPBottom');
        wrapPInput = document.createElement('input');
        wrapPInput.classList.add('MPwrapPTopInput_bottom');
        wrapPInput.setAttribute('name', 'padding_bottom');
        wrapPInput.setAttribute('type', 'number');
        // wrapPInput.setAttribute('min', '0');
        wrapPInput.setAttribute('max', '96');
        wrapPInput.setAttribute('step', '0.5');
        wrapPInput.setAttribute('value', this.data.Padding.paddingB ? this.data.Padding.paddingB : '');
        wrapP3.appendChild(wrapPInput);

        wrap.appendChild(wrapP);
        wrap.appendChild(wrapP2);
        wrap.appendChild(wrapP3);
        wrapper.appendChild(wrap);
        let self = this;
        wrapper.querySelector('[name=padding]').addEventListener('input', function(e){
            let v = e.target.value;
            wrapper.querySelector('[name=padding_top]').value = v;
            wrapper.querySelector('[name=padding_bottom]').value = v;
            wrapper.querySelector('[name=padding_left]').value = v;
            wrapper.querySelector('[name=padding_right]').value = v;
            self.save()
        });
        wrapper.querySelector('[name=padding_top]').addEventListener('input', function(e){
            self.save()
        })
        wrapper.querySelector('[name=padding_bottom]').addEventListener('input', function(e){
            self.save()
        })
        wrapper.querySelector('[name=padding_left]').addEventListener('input', function(e){
            self.save()
        }) 
        wrapper.querySelector('[name=padding_right]').addEventListener('input', function(e){
            self.save()
        })
        this.wrapper = wrapper;
        // this.alignmentSettings.map(align => {
        //     const button = document.createElement('button');
        //     button.classList.add(this.api.styles.settingsButton);
        //     // button.classList.add('cdx-settings-button-header');
        //     button.innerHTML = align.icon;
        //     button.type = 'button';
          
        //     button.classList.toggle(this.api.styles.settingsButtonActive, align.prop.className === this.data.Padding);
        //     wrapper.appendChild(button);
        //     return button
        // }).forEach((element, index, elements) => {
           
        //     element.addEventListener('click', () => {
        //         this.data = {
        //             Padding: this.alignmentSettings[index].prop.className
        //         }
        //         elements.forEach((el, i) => {
        //             const {className} = this.alignmentSettings[i].prop;
        //             el.classList.toggle(this.api.styles.settingsButtonActive, className === this.data.Padding);
        //             el.setAttribute('title', this.alignmentSettings[i].title);
        //             //toggle alignment style class for block
        //             if (className.includes(' ')){
        //                 let values = className.split(' ');
        //                 for( var i = 0; i < values.length; i++){
        //                     this.wrapper.classList.toggle(`${values[i]}`, className === this.data.Padding)
        //                 }
        //             }else{

        //                 this.wrapper.classList.toggle(`${className}`, className === this.data.Padding)
        //             }
        //             // this.wrapper.classList.toggle(this._CSS.alignment[name], name === this.data.alignment)
        //         });
        //     });
        // });
        // wrapper.appendChild(wrapperColorsText);
        this.wrp = wrapper;
        return this.wrp;
    }
    /**
     * save
     * @returns {*}
     */
    save() {
        if (this.wrp){
            let toolsContent = this.wrp;
            let p = toolsContent.querySelector('[name=padding]').value;
            let pt = toolsContent.querySelector('[name=padding_top]').value;
            let pb = toolsContent.querySelector('[name=padding_bottom]').value;
            let pl = toolsContent.querySelector('[name=padding_left]').value;
            let pr = toolsContent.querySelector('[name=padding_right]').value;
            let pClass = '';
            if (p == ''){
                if (pt == pb){     
                    if ((pt !== '') && (pb !== '')){
                        pClass += ` py-${pt}`;
                    }
                    
                }else{
                    if(pt !== ''){
                        pClass += ` pt-${pt}`;
                    }
                    if(pb !== ''){
                        pClass += ` pb-${pb}`;
                    }
                }

                if (pl == pr){
                    if ((pl !== '') && (pr !== '')){
                        pClass += ` px-${pl}`;
                    }
                    
                }else{
                    if(pl !== ''){
                        pClass += ` pt-${pl}`;
                    }
                    if(pr !== ''){
                        pClass += ` pb-${pr}`;
                    }                
                }
                this.data.Padding.padding = '';
            }else{
                this.data.Padding.padding = p;
                pClass = `p-${p}`;
            }
            this.data.Padding.paddingY = pt;
            this.data.Padding.paddingT = pt;
            this.data.Padding.paddingB = pb;
            this.data.Padding.paddingX = pl;
            this.data.Padding.paddingL = pl;
            this.data.Padding.paddingR = pr;
            
            this.data.Padding.paddingClass =  pClass;
            
        }
        return this.data;
    }

    make(tagName, classNames = null, attributes = {}) {
        const el = document.createElement(tagName);
      
        if (Array.isArray(classNames)) {
          el.classList.add(...classNames);
        } else if (classNames) {
          el.classList.add(classNames);
        }
      
        for (const attrName in attributes) {
          el[attrName] = attributes[attrName];
        }
        return el;
    }
}

module.exports = PaddingTune;
