// @silver0542/directus-extension-cma-components-editor/src/custom-plugins/tune-textStyle.js


class textStyleTune {

    /**
     * Default alignment
     *
     * @public
     * @returns {string}
     */
    static get DEFAULT_ALIGNMENT() {
        return 'text-base';
    }

    static get isTune() {
        return true;
    }

    getAlignment(){
        if(!!this.settings?.blocks && this.settings.blocks.hasOwnProperty(this.block.name)){
            return this.settings.blocks[this.block.name]
        }
        if(!!this.settings?.default){
            return this.settings.default
        }
        return textStyleTune.DEFAULT_ALIGNMENT
    }
    /**
     *
     * @param api
     * @param data 既に設定されているデータ
     * @param settings tuneに設定項目
     * @param block tuneに設定されてるblock
     */
    constructor({ api, data, config, block}) {
        this.api = api;
        this.block = block;
        /**
         config:{
            default: "right",
            blocks: {
              header: 'center',
              list: 'right'
            }
          },
         */
        this.settings = config;
        this.data = data || { textStyle: this.getAlignment() }
        this.alignmentSettings = [
            {
				name: 'XS',
				title: 'XS',
				type: 'header_container',
				prop: { type: 'text', className:'text-xs'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M8.00186 7.8L7.40186 8.916L6.81386 7.8H5.36786L6.68186 9.858L5.21786 12H6.66386L7.38986 10.71L8.09186 12H9.54386L8.10986 9.858L9.45386 7.8H8.00186ZM11.8875 10.464L11.0955 11.04C11.1875 11.248 11.3235 11.438 11.5035 11.61C11.6835 11.778 11.8955 11.91 12.1395 12.006C12.3875 12.106 12.6555 12.156 12.9435 12.156C13.1475 12.156 13.3475 12.126 13.5435 12.066C13.7395 12.01 13.9135 11.926 14.0655 11.814C14.2215 11.698 14.3435 11.554 14.4315 11.382C14.5235 11.21 14.5695 11.012 14.5695 10.788C14.5695 10.62 14.5435 10.468 14.4915 10.332C14.4395 10.196 14.3655 10.076 14.2695 9.972C14.1735 9.864 14.0595 9.772 13.9275 9.696C13.7995 9.616 13.6575 9.548 13.5015 9.492C13.3215 9.428 13.1615 9.37 13.0215 9.318C12.8855 9.262 12.7775 9.204 12.6975 9.144C12.6215 9.08 12.5835 9.002 12.5835 8.91C12.5835 8.83 12.6175 8.766 12.6855 8.718C12.7535 8.67 12.8455 8.646 12.9615 8.646C13.0535 8.646 13.1435 8.664 13.2315 8.7C13.3195 8.736 13.4015 8.788 13.4775 8.856C13.5575 8.92 13.6275 9.002 13.6875 9.102L14.5635 8.616C14.4875 8.472 14.3795 8.33 14.2395 8.19C14.0995 8.046 13.9215 7.928 13.7055 7.836C13.4895 7.74 13.2335 7.692 12.9375 7.692C12.6535 7.692 12.3955 7.74 12.1635 7.836C11.9315 7.932 11.7455 8.072 11.6055 8.256C11.4655 8.436 11.3955 8.658 11.3955 8.922C11.3955 9.13 11.4335 9.31 11.5095 9.462C11.5895 9.614 11.6875 9.742 11.8035 9.846C11.9235 9.95 12.0475 10.036 12.1755 10.104C12.3035 10.168 12.4195 10.22 12.5235 10.26C12.7355 10.34 12.8995 10.408 13.0155 10.464C13.1355 10.52 13.2195 10.576 13.2675 10.632C13.3155 10.684 13.3395 10.75 13.3395 10.83C13.3395 10.942 13.2975 11.028 13.2135 11.088C13.1295 11.144 13.0235 11.172 12.8955 11.172C12.7635 11.172 12.6355 11.144 12.5115 11.088C12.3875 11.032 12.2735 10.952 12.1695 10.848C12.0655 10.74 11.9715 10.612 11.8875 10.464Z" fill="black"/>
                </svg>                                                            
				`,
			},
			{
				name: 'SM',
				title: 'SM',
				type: 'header_container',
				prop: { type: 'text', className: 'text-sm'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M5.61857 10.464L4.82657 11.04C4.91857 11.248 5.05457 11.438 5.23457 11.61C5.41457 11.778 5.62657 11.91 5.87057 12.006C6.11857 12.106 6.38657 12.156 6.67457 12.156C6.87857 12.156 7.07857 12.126 7.27457 12.066C7.47057 12.01 7.64457 11.926 7.79657 11.814C7.95257 11.698 8.07457 11.554 8.16257 11.382C8.25457 11.21 8.30057 11.012 8.30057 10.788C8.30057 10.62 8.27457 10.468 8.22257 10.332C8.17057 10.196 8.09657 10.076 8.00057 9.972C7.90457 9.864 7.79057 9.772 7.65857 9.696C7.53057 9.616 7.38857 9.548 7.23257 9.492C7.05257 9.428 6.89257 9.37 6.75257 9.318C6.61657 9.262 6.50857 9.204 6.42857 9.144C6.35257 9.08 6.31457 9.002 6.31457 8.91C6.31457 8.83 6.34857 8.766 6.41657 8.718C6.48457 8.67 6.57657 8.646 6.69257 8.646C6.78457 8.646 6.87457 8.664 6.96257 8.7C7.05057 8.736 7.13257 8.788 7.20857 8.856C7.28857 8.92 7.35857 9.002 7.41857 9.102L8.29457 8.616C8.21857 8.472 8.11057 8.33 7.97057 8.19C7.83057 8.046 7.65257 7.928 7.43657 7.836C7.22057 7.74 6.96457 7.692 6.66857 7.692C6.38457 7.692 6.12657 7.74 5.89457 7.836C5.66257 7.932 5.47657 8.072 5.33657 8.256C5.19657 8.436 5.12657 8.658 5.12657 8.922C5.12657 9.13 5.16457 9.31 5.24057 9.462C5.32057 9.614 5.41857 9.742 5.53457 9.846C5.65457 9.95 5.77857 10.036 5.90657 10.104C6.03457 10.168 6.15057 10.22 6.25457 10.26C6.46657 10.34 6.63057 10.408 6.74657 10.464C6.86657 10.52 6.95057 10.576 6.99857 10.632C7.04657 10.684 7.07057 10.75 7.07057 10.83C7.07057 10.942 7.02857 11.028 6.94457 11.088C6.86057 11.144 6.75457 11.172 6.62657 11.172C6.49457 11.172 6.36657 11.144 6.24257 11.088C6.11857 11.032 6.00457 10.952 5.90057 10.848C5.79657 10.74 5.70257 10.612 5.61857 10.464ZM11.5372 10.068L12.6172 11.61L13.7032 10.068L13.9372 12H15.1732L14.4472 7.59L12.6172 10.008L10.7932 7.59L10.0672 12H11.3032L11.5372 10.068Z" fill="black"/>
                </svg>
                                                                        
				`,
			},
            {
				name: 'BASE',
				title: 'BASE',
				type: 'header_container',
				prop: { type: 'text', className: 'text-base'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M5.84384 9.632H6.39184C6.59184 9.632 6.76918 9.60533 6.92384 9.552C7.08118 9.496 7.20518 9.41467 7.29584 9.308C7.38651 9.19867 7.43184 9.068 7.43184 8.916C7.43184 8.756 7.38651 8.624 7.29584 8.52C7.20518 8.41333 7.08118 8.33333 6.92384 8.28C6.76918 8.22667 6.59184 8.2 6.39184 8.2H5.26784V11H6.46384C6.69851 11 6.89984 10.9747 7.06784 10.924C7.23584 10.8707 7.36518 10.7867 7.45584 10.672C7.54651 10.5573 7.59184 10.4067 7.59184 10.22C7.59184 10.076 7.55851 9.95333 7.49184 9.852C7.42784 9.748 7.34118 9.664 7.23184 9.6C7.12251 9.53333 7.00118 9.48533 6.86784 9.456C6.73451 9.424 6.59984 9.408 6.46384 9.408H5.84384V9.832H6.36784C6.42651 9.832 6.47984 9.836 6.52784 9.844C6.57851 9.852 6.62251 9.86667 6.65984 9.888C6.69984 9.90933 6.73051 9.93733 6.75184 9.972C6.77318 10.0067 6.78384 10.0507 6.78384 10.104C6.78384 10.1867 6.76518 10.2507 6.72784 10.296C6.69051 10.3387 6.63984 10.368 6.57584 10.384C6.51451 10.4 6.44518 10.408 6.36784 10.408H6.04384V8.792H6.29584C6.40518 8.792 6.48651 8.812 6.53984 8.852C6.59584 8.892 6.62384 8.95467 6.62384 9.04C6.62384 9.096 6.61051 9.144 6.58384 9.184C6.55984 9.224 6.52384 9.25467 6.47584 9.276C6.42784 9.29733 6.36784 9.308 6.29584 9.308H5.84384V9.632ZM8.2225 10.548H9.7025L9.6825 9.976H8.2465L8.2225 10.548ZM8.9545 9.348L9.3265 10.188L9.2825 10.388L9.6025 11H10.4825L8.9545 8.012L7.4305 11H8.3105L8.6345 10.36L8.5865 10.188L8.9545 9.348ZM10.8658 9.976L10.3378 10.36C10.3992 10.4987 10.4898 10.6253 10.6098 10.74C10.7298 10.852 10.8712 10.94 11.0338 11.004C11.1992 11.0707 11.3778 11.104 11.5698 11.104C11.7058 11.104 11.8392 11.084 11.9698 11.044C12.1005 11.0067 12.2165 10.9507 12.3178 10.876C12.4218 10.7987 12.5032 10.7027 12.5618 10.588C12.6232 10.4733 12.6538 10.3413 12.6538 10.192C12.6538 10.08 12.6365 9.97867 12.6018 9.888C12.5672 9.79733 12.5178 9.71733 12.4538 9.648C12.3898 9.576 12.3138 9.51467 12.2258 9.464C12.1405 9.41067 12.0458 9.36533 11.9418 9.328C11.8218 9.28533 11.7152 9.24667 11.6218 9.212C11.5312 9.17467 11.4592 9.136 11.4058 9.096C11.3552 9.05333 11.3298 9.00133 11.3298 8.94C11.3298 8.88667 11.3525 8.844 11.3978 8.812C11.4432 8.78 11.5045 8.764 11.5818 8.764C11.6432 8.764 11.7032 8.776 11.7618 8.8C11.8205 8.824 11.8752 8.85867 11.9258 8.904C11.9792 8.94667 12.0258 9.00133 12.0658 9.068L12.6498 8.744C12.5992 8.648 12.5272 8.55333 12.4338 8.46C12.3405 8.364 12.2218 8.28533 12.0778 8.224C11.9338 8.16 11.7632 8.128 11.5658 8.128C11.3765 8.128 11.2045 8.16 11.0498 8.224C10.8952 8.288 10.7712 8.38133 10.6778 8.504C10.5845 8.624 10.5378 8.772 10.5378 8.948C10.5378 9.08667 10.5632 9.20667 10.6138 9.308C10.6672 9.40933 10.7325 9.49467 10.8098 9.564C10.8898 9.63333 10.9725 9.69067 11.0578 9.736C11.1432 9.77867 11.2205 9.81333 11.2898 9.84C11.4312 9.89333 11.5405 9.93867 11.6178 9.976C11.6978 10.0133 11.7538 10.0507 11.7858 10.088C11.8178 10.1227 11.8338 10.1667 11.8338 10.22C11.8338 10.2947 11.8058 10.352 11.7498 10.392C11.6938 10.4293 11.6232 10.448 11.5378 10.448C11.4498 10.448 11.3645 10.4293 11.2818 10.392C11.1992 10.3547 11.1232 10.3013 11.0538 10.232C10.9845 10.16 10.9218 10.0747 10.8658 9.976ZM13.3596 11H14.7716V10.372H13.3596V11ZM13.3596 8.828H14.7716V8.2H13.3596V8.828ZM13.3596 9.848H14.6916V9.24H13.3596V9.848ZM12.8316 8.2V11H13.5756V8.2H12.8316Z" fill="black"/>
                </svg>                                                                
				`,
			},
            {
				name: 'LG',
				title: 'LG',
				type: 'header_container',
				prop: { type: 'text', className: 'text-lg'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M5.71696 7.8V12H8.57896V11.01H6.88096V7.8H5.71696ZM12.308 10.518H13.346C13.31 10.61 13.262 10.696 13.202 10.776C13.146 10.856 13.078 10.926 12.998 10.986C12.918 11.042 12.824 11.086 12.716 11.118C12.608 11.146 12.484 11.16 12.344 11.16C12.112 11.16 11.908 11.11 11.732 11.01C11.556 10.91 11.418 10.766 11.318 10.578C11.218 10.386 11.168 10.16 11.168 9.9C11.168 9.636 11.22 9.414 11.324 9.234C11.432 9.054 11.576 8.918 11.756 8.826C11.936 8.734 12.132 8.688 12.344 8.688C12.568 8.688 12.75 8.74 12.89 8.844C13.03 8.948 13.142 9.076 13.226 9.228L14.288 8.718C14.152 8.522 13.992 8.348 13.808 8.196C13.624 8.04 13.41 7.918 13.166 7.83C12.926 7.738 12.652 7.692 12.344 7.692C12.004 7.692 11.688 7.744 11.396 7.848C11.104 7.948 10.85 8.096 10.634 8.292C10.418 8.484 10.25 8.716 10.13 8.988C10.01 9.26 9.95001 9.564 9.95001 9.9C9.95001 10.236 10.008 10.544 10.124 10.824C10.24 11.1 10.402 11.338 10.61 11.538C10.822 11.734 11.07 11.886 11.354 11.994C11.642 12.102 11.956 12.156 12.296 12.156C12.664 12.156 12.99 12.092 13.274 11.964C13.558 11.832 13.794 11.652 13.982 11.424C14.17 11.196 14.308 10.93 14.396 10.626C14.484 10.322 14.514 9.998 14.486 9.654H12.308V10.518Z" fill="black"/>
                </svg>                                                                              
				`,
			},
            {
				name: 'XL',
				title: 'XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M8.29483 7.8L7.69483 8.916L7.10683 7.8H5.66083L6.97483 9.858L5.51083 12H6.95683L7.68283 10.71L8.38483 12H9.83683L8.40283 9.858L9.74683 7.8H8.29483ZM11.6285 7.8V12H14.4905V11.01H12.7925V7.8H11.6285Z" fill="black"/>
                </svg>                                                                                            
				`,
			},
            {
				name: '2XL',
				title: '2XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-2xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M4.36891 12H8.10691V10.938H6.49291L7.36291 10.206C7.55091 10.054 7.70091 9.886 7.81291 9.702C7.92491 9.518 7.98091 9.312 7.98091 9.084C7.98091 8.912 7.94491 8.746 7.87291 8.586C7.80091 8.422 7.69291 8.274 7.54891 8.142C7.40891 8.01 7.23491 7.906 7.02691 7.83C6.81891 7.75 6.58091 7.71 6.31291 7.71C5.92891 7.71 5.60491 7.788 5.34091 7.944C5.08091 8.096 4.88491 8.3 4.75291 8.556C4.62491 8.808 4.56091 9.084 4.56091 9.384H5.70091C5.70091 9.252 5.72291 9.134 5.76691 9.03C5.81491 8.926 5.88291 8.846 5.97091 8.79C6.05891 8.73 6.15891 8.7 6.27091 8.7C6.35891 8.7 6.43291 8.714 6.49291 8.742C6.55691 8.766 6.61091 8.8 6.65491 8.844C6.69891 8.884 6.73091 8.932 6.75091 8.988C6.77091 9.044 6.78091 9.102 6.78091 9.162C6.78091 9.27 6.75691 9.374 6.70891 9.474C6.66491 9.57 6.59891 9.668 6.51091 9.768C6.42691 9.868 6.32691 9.984 6.21091 10.116L4.36891 12ZM11.0148 7.8L10.4148 8.916L9.82675 7.8H8.38075L9.69475 9.858L8.23075 12H9.67675L10.4028 10.71L11.1048 12H12.5568L11.1228 9.858L12.4668 7.8H11.0148ZM12.8784 7.8V12H15.7404V11.01H14.0424V7.8H12.8784Z" fill="black"/>
                </svg>                                                                                                          
				`,
			},
            {
				name: '3XL',
				title: '3XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-3xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M6.09604 9.978C6.44804 9.978 6.76804 9.942 7.05604 9.87C7.34404 9.794 7.57204 9.676 7.74004 9.516C7.91204 9.356 7.99804 9.148 7.99804 8.892C7.99804 8.672 7.93604 8.474 7.81204 8.298C7.69204 8.118 7.51404 7.976 7.27804 7.872C7.04604 7.768 6.76204 7.716 6.42604 7.716C6.11004 7.716 5.83404 7.778 5.59804 7.902C5.36204 8.022 5.17804 8.186 5.04604 8.394C4.91804 8.598 4.85404 8.828 4.85404 9.084H5.92204C5.92204 8.948 5.96804 8.838 6.06004 8.754C6.15604 8.67 6.27804 8.628 6.42604 8.628C6.52204 8.628 6.60004 8.642 6.66004 8.67C6.72404 8.698 6.77204 8.738 6.80404 8.79C6.83604 8.842 6.85204 8.902 6.85204 8.97C6.85204 9.038 6.83404 9.104 6.79804 9.168C6.76204 9.228 6.71004 9.282 6.64204 9.33C6.57404 9.378 6.49404 9.416 6.40204 9.444C6.31004 9.472 6.20804 9.486 6.09604 9.486V9.978ZM6.37204 12.108C6.74404 12.108 7.06004 12.054 7.32004 11.946C7.58004 11.838 7.77804 11.688 7.91404 11.496C8.05004 11.304 8.11804 11.086 8.11804 10.842C8.11804 10.626 8.06804 10.442 7.96804 10.29C7.87204 10.138 7.73404 10.012 7.55404 9.912C7.37404 9.812 7.16004 9.74 6.91204 9.696C6.66804 9.652 6.39604 9.63 6.09604 9.63V10.194C6.25604 10.194 6.39004 10.208 6.49804 10.236C6.61004 10.264 6.70004 10.304 6.76804 10.356C6.84004 10.404 6.89204 10.462 6.92404 10.53C6.95604 10.594 6.97204 10.662 6.97204 10.734C6.97204 10.842 6.94804 10.932 6.90004 11.004C6.85204 11.072 6.78604 11.12 6.70204 11.148C6.62204 11.176 6.52804 11.19 6.42004 11.19C6.32004 11.19 6.22204 11.17 6.12604 11.13C6.03404 11.09 5.96004 11.026 5.90404 10.938C5.84804 10.85 5.82004 10.738 5.82004 10.602H4.67404C4.67404 10.794 4.71204 10.98 4.78804 11.16C4.86404 11.34 4.97404 11.502 5.11804 11.646C5.26604 11.79 5.44404 11.904 5.65204 11.988C5.86404 12.068 6.10404 12.108 6.37204 12.108ZM10.8536 7.8L10.2536 8.916L9.66562 7.8H8.21962L9.53362 9.858L8.06962 12H9.51562L10.2416 10.71L10.9436 12H12.3956L10.9616 9.858L12.3056 7.8H10.8536ZM12.7173 7.8V12H15.5793V11.01H13.8813V7.8H12.7173Z" fill="black"/>
                </svg>
                                                                                                                        
				`,
			},


            {
				name: '4XL',
				title: '4XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-4xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M4.16856 11.28H8.21256V10.296H7.35456L7.25256 10.47H5.66256L6.48456 9.372V10.878L6.34056 10.944V12H7.55256V7.8H6.64056L4.16856 11.28ZM11.1671 7.8L10.5671 8.916L9.97909 7.8H8.53309L9.84709 9.858L8.38309 12H9.82909L10.5551 10.71L11.2571 12H12.7091L11.2751 9.858L12.6191 7.8H11.1671ZM13.0308 7.8V12H15.8928V11.01H14.1948V7.8H13.0308Z" fill="black"/>
                </svg>                                                                                                             
				`,
			},
             {
				name: '5XL',
				title: '5XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-5xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M8.06099 10.668C8.06099 10.372 7.99899 10.124 7.87499 9.924C7.75099 9.72 7.59099 9.566 7.39499 9.462C7.19899 9.358 6.99299 9.306 6.77699 9.306C6.63699 9.306 6.50299 9.316 6.37499 9.336C6.24699 9.356 6.12699 9.394 6.01499 9.45L6.23699 8.808H7.95299V7.8H5.39099L4.79699 10.278C4.94099 10.214 5.08499 10.166 5.22899 10.134C5.37699 10.102 5.51899 10.08 5.65499 10.068C5.79099 10.056 5.90899 10.05 6.00899 10.05C6.13299 10.05 6.25299 10.066 6.36899 10.098C6.48899 10.126 6.58699 10.174 6.66299 10.242C6.74299 10.31 6.78299 10.404 6.78299 10.524C6.78299 10.64 6.75099 10.736 6.68699 10.812C6.62299 10.888 6.53299 10.942 6.41699 10.974C6.30499 11.006 6.16899 11.022 6.00899 11.022C5.92099 11.022 5.81899 11.01 5.70299 10.986C5.58699 10.958 5.46299 10.916 5.33099 10.86C5.20299 10.804 5.07499 10.732 4.94699 10.644L4.56299 11.64C4.70299 11.72 4.85699 11.794 5.02499 11.862C5.19699 11.926 5.38899 11.978 5.60099 12.018C5.81299 12.062 6.04699 12.084 6.30299 12.084C6.50699 12.084 6.71099 12.056 6.91499 12C7.12299 11.948 7.31299 11.866 7.48499 11.754C7.65699 11.638 7.79499 11.49 7.89899 11.31C8.00699 11.13 8.06099 10.916 8.06099 10.668ZM10.9767 7.8L10.3767 8.916L9.78866 7.8H8.34266L9.65666 9.858L8.19266 12H9.63866L10.3647 10.71L11.0667 12H12.5187L11.0847 9.858L12.4287 7.8H10.9767ZM12.8403 7.8V12H15.7023V11.01H14.0043V7.8H12.8403Z" fill="black"/>
                </svg>
                                                                                                                             
				`,
			},
             {
				name: '6XL',
				title: '6XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-6xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M5.65403 10.5C5.65403 10.372 5.68203 10.26 5.73803 10.164C5.79403 10.068 5.87003 9.994 5.96603 9.942C6.06603 9.89 6.17603 9.864 6.29603 9.864C6.41603 9.864 6.52403 9.89 6.62003 9.942C6.71603 9.994 6.79203 10.068 6.84803 10.164C6.90403 10.26 6.93203 10.372 6.93203 10.5C6.93203 10.628 6.90403 10.738 6.84803 10.83C6.79203 10.922 6.71603 10.994 6.62003 11.046C6.52403 11.098 6.41603 11.124 6.29603 11.124C6.17603 11.124 6.06603 11.098 5.96603 11.046C5.87003 10.994 5.79403 10.922 5.73803 10.83C5.68203 10.738 5.65403 10.628 5.65403 10.5ZM6.20603 7.824L5.05403 9.24C4.90203 9.428 4.77203 9.624 4.66403 9.828C4.55603 10.032 4.50203 10.256 4.50203 10.5C4.50203 10.86 4.58403 11.16 4.74803 11.4C4.91203 11.636 5.12803 11.814 5.39603 11.934C5.66803 12.05 5.96803 12.108 6.29603 12.108C6.62003 12.108 6.91803 12.05 7.19003 11.934C7.46203 11.814 7.67803 11.636 7.83803 11.4C8.00203 11.16 8.08403 10.86 8.08403 10.5C8.08403 10.28 8.04403 10.086 7.96403 9.918C7.88803 9.75 7.78603 9.608 7.65803 9.492C7.53003 9.372 7.38803 9.282 7.23203 9.222C7.07603 9.162 6.92203 9.132 6.77003 9.132C6.61403 9.132 6.48603 9.15 6.38603 9.186C6.28603 9.218 6.17403 9.29 6.05003 9.402L6.30803 9.276L7.68203 7.824H6.20603ZM11.0616 7.8L10.4616 8.916L9.87363 7.8H8.42763L9.74163 9.858L8.27763 12H9.72363L10.4496 10.71L11.1516 12H12.6036L11.1696 9.858L12.5136 7.8H11.0616ZM12.9253 7.8V12H15.7873V11.01H14.0893V7.8H12.9253Z" fill="black"/>
                </svg>                                                                                                                       
				`,
			},





             {
				name: '7XL',
				title: '7XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-7xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M4.5959 8.886H6.4619L4.8719 12H6.1319L8.1659 7.8H4.5959V8.886ZM10.8478 7.8L10.2478 8.916L9.65976 7.8H8.21376L9.52776 9.858L8.06376 12H9.50976L10.2358 10.71L10.9378 12H12.3898L10.9558 9.858L12.2998 7.8H10.8478ZM12.7114 7.8V12H15.5734V11.01H13.8754V7.8H12.7114Z" fill="black"/>
                </svg>
                                                                                                                                     
				`,
			},
             {
				name: '8XL',
				title: '8XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-8xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M4.83967 8.814C4.83967 9.014 4.88367 9.188 4.97167 9.336C5.06367 9.48 5.18367 9.6 5.33167 9.696C5.47967 9.788 5.63967 9.856 5.81167 9.9C5.98767 9.944 6.16167 9.966 6.33367 9.966C6.50567 9.966 6.67967 9.944 6.85567 9.9C7.03167 9.856 7.19367 9.788 7.34167 9.696C7.48967 9.6 7.60767 9.48 7.69567 9.336C7.78767 9.188 7.83367 9.014 7.83367 8.814C7.83367 8.566 7.76367 8.36 7.62367 8.196C7.48767 8.028 7.30567 7.902 7.07767 7.818C6.85367 7.734 6.60567 7.692 6.33367 7.692C6.06567 7.692 5.81767 7.734 5.58967 7.818C5.36567 7.902 5.18367 8.028 5.04367 8.196C4.90767 8.36 4.83967 8.566 4.83967 8.814ZM5.93167 8.922C5.93167 8.838 5.94767 8.764 5.97967 8.7C6.01567 8.636 6.06367 8.588 6.12367 8.556C6.18767 8.52 6.25767 8.502 6.33367 8.502C6.41767 8.502 6.48967 8.52 6.54967 8.556C6.60967 8.588 6.65567 8.636 6.68767 8.7C6.72367 8.764 6.74167 8.838 6.74167 8.922C6.74167 9.006 6.72367 9.08 6.68767 9.144C6.65567 9.204 6.60967 9.254 6.54967 9.294C6.48967 9.33 6.41767 9.348 6.33367 9.348C6.25767 9.348 6.18767 9.33 6.12367 9.294C6.06367 9.254 6.01567 9.204 5.97967 9.144C5.94767 9.08 5.93167 9.006 5.93167 8.922ZM4.68967 10.86C4.68967 11.06 4.73367 11.238 4.82167 11.394C4.91367 11.55 5.03767 11.68 5.19367 11.784C5.34967 11.888 5.52567 11.968 5.72167 12.024C5.92167 12.08 6.12567 12.108 6.33367 12.108C6.54567 12.108 6.74967 12.08 6.94567 12.024C7.14167 11.968 7.31767 11.888 7.47367 11.784C7.63367 11.68 7.75767 11.55 7.84567 11.394C7.93767 11.238 7.98367 11.06 7.98367 10.86C7.98367 10.628 7.93167 10.434 7.82767 10.278C7.72367 10.118 7.58767 9.99 7.41967 9.894C7.25167 9.794 7.07167 9.722 6.87967 9.678C6.68767 9.634 6.50567 9.612 6.33367 9.612C6.16167 9.612 5.97967 9.634 5.78767 9.678C5.59967 9.722 5.42167 9.794 5.25367 9.894C5.08567 9.99 4.94967 10.118 4.84567 10.278C4.74167 10.434 4.68967 10.628 4.68967 10.86ZM5.85367 10.68C5.85367 10.592 5.87367 10.512 5.91367 10.44C5.95767 10.368 6.01567 10.312 6.08767 10.272C6.16367 10.228 6.24567 10.206 6.33367 10.206C6.42567 10.206 6.50767 10.228 6.57967 10.272C6.65567 10.312 6.71367 10.368 6.75367 10.44C6.79767 10.512 6.81967 10.592 6.81967 10.68C6.81967 10.788 6.79767 10.878 6.75367 10.95C6.71367 11.022 6.65567 11.078 6.57967 11.118C6.50767 11.154 6.42567 11.172 6.33367 11.172C6.24567 11.172 6.16367 11.154 6.08767 11.118C6.01567 11.078 5.95767 11.022 5.91367 10.95C5.87367 10.878 5.85367 10.788 5.85367 10.68ZM10.88 7.8L10.28 8.916L9.69198 7.8H8.24598L9.55998 9.858L8.09598 12H9.54198L10.268 10.71L10.97 12H12.422L10.988 9.858L12.332 7.8H10.88ZM12.7436 7.8V12H15.6056V11.01H13.9076V7.8H12.7436Z" fill="black"/>
                </svg>                
				`,
			},
             {
				name: '9XL',
				title: '9XL',
				type: 'header_container',
				prop: { type: 'text', className: 'text-9xl'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
                <rect width="20" height="20" rx="5" fill="white"/>
                <path d="M7.02285 9.3C7.02285 9.428 6.99485 9.54 6.93885 9.636C6.88285 9.728 6.80685 9.802 6.71085 9.858C6.61485 9.91 6.50685 9.936 6.38685 9.936C6.26685 9.936 6.15685 9.91 6.05685 9.858C5.96085 9.802 5.88485 9.728 5.82885 9.636C5.77285 9.54 5.74485 9.428 5.74485 9.3C5.74485 9.172 5.77285 9.062 5.82885 8.97C5.88485 8.878 5.96085 8.806 6.05685 8.754C6.15685 8.702 6.26685 8.676 6.38685 8.676C6.50685 8.676 6.61485 8.702 6.71085 8.754C6.80685 8.806 6.88285 8.878 6.93885 8.97C6.99485 9.062 7.02285 9.172 7.02285 9.3ZM6.47085 11.976L7.62285 10.56C7.77485 10.372 7.90485 10.176 8.01285 9.972C8.12085 9.768 8.17485 9.544 8.17485 9.3C8.17485 8.94 8.09285 8.642 7.92885 8.406C7.76885 8.166 7.55285 7.988 7.28085 7.872C7.00885 7.752 6.71085 7.692 6.38685 7.692C6.05885 7.692 5.75885 7.752 5.48685 7.872C5.21885 7.988 5.00285 8.166 4.83885 8.406C4.67485 8.642 4.59285 8.94 4.59285 9.3C4.59285 9.52 4.63085 9.714 4.70685 9.882C4.78685 10.05 4.89085 10.194 5.01885 10.314C5.15085 10.43 5.29285 10.518 5.44485 10.578C5.60085 10.638 5.75485 10.668 5.90685 10.668C6.06285 10.668 6.19085 10.65 6.29085 10.614C6.39485 10.578 6.50685 10.506 6.62685 10.398L6.36885 10.524L4.99485 11.976H6.47085ZM10.9708 7.8L10.3708 8.916L9.7828 7.8H8.3368L9.6508 9.858L8.1868 12H9.6328L10.3588 10.71L11.0608 12H12.5128L11.0788 9.858L12.4228 7.8H10.9708ZM12.8345 7.8V12H15.6965V11.01H13.9985V7.8H12.8345Z" fill="black"/>
                </svg>                                                                                                                                       
				`,
			},
        ];

    }

    /**
     * block自体をwrapしてくれる
     * constructorで与えられたalignmentを代入しようとすると、holderが確定してなく
     * renderでやろうとすると、tuneを表示時に処理が走る
     * @param blockContent
     */
    wrap(blockContent) {
        this.wrapper = this.make("div");
        this.wrapper.classList.toggle(`${this.data.textStyle}`)
        this.wrapper.append(blockContent)
        return this.wrapper
    }

    /**
     * rendering block tune
     * @returns {*}
     */
    render() {
        const wrapper = this.make("div");

		wrapper.classList.add('cdx-settings-button-header-wrapper');

        this.alignmentSettings.map(align => {
            const button = document.createElement('button');
            button.classList.add(this.api.styles.settingsButton);
            // button.classList.add('cdx-settings-button-header');
            button.innerHTML = align.icon;
            button.type = 'button';
          
            button.classList.toggle(this.api.styles.settingsButtonActive, align.prop.className === this.data.textStyle);
            wrapper.appendChild(button);
            return button
        }).forEach((element, index, elements) => {
           
            element.addEventListener('click', () => {
                this.data = {
                    textStyle: this.alignmentSettings[index].prop.className
                }
                elements.forEach((el, i) => {
                    const {className} = this.alignmentSettings[i].prop;
                    el.classList.toggle(this.api.styles.settingsButtonActive, className === this.data.textStyle);
                    el.setAttribute('title', this.alignmentSettings[i].title);
                    //toggle alignment style class for block
                    if (className.includes(' ')){
                        let values = className.split(' ');
                        for( var i = 0; i < values.length; i++){
                            this.wrapper.classList.toggle(`${values[i]}`, className === this.data.textStyle)
                        }
                    }else{

                        this.wrapper.classList.toggle(`${className}`, className === this.data.textStyle)
                    }
                    // this.wrapper.classList.toggle(this._CSS.alignment[name], name === this.data.alignment)
                });
            });
        });
        // wrapper.appendChild(wrapperColorsText);
        return wrapper;
    }
    /**
     * save
     * @returns {*}
     */
    save() {
        return this.data;
    }

    make(tagName, classNames = null, attributes = {}) {
        const el = document.createElement(tagName);
      
        if (Array.isArray(classNames)) {
          el.classList.add(...classNames);
        } else if (classNames) {
          el.classList.add(classNames);
        }
      
        for (const attrName in attributes) {
          el[attrName] = attributes[attrName];
        }
        return el;
    }
}

module.exports = textStyleTune;
