// @silver0542/directus-extension-cma-components-editor/src/custom-plugins/tune-background.js


class containerBgTune {

    /**
     * Default alignment
     *
     * @public
     * @returns {string}
     */
    static get DEFAULT_ALIGNMENT() {
        return 'white';
    }

    static get isTune() {
        return true;
    }

    getAlignment(){
        if(!!this.settings?.blocks && this.settings.blocks.hasOwnProperty(this.block.name)){
            return this.settings.blocks[this.block.name]
        }
        if(!!this.settings?.default){
            return this.settings.default
        }
        return containerBgTune.DEFAULT_ALIGNMENT
    }
    /**
     *
     * @param api
     * @param data 既に設定されているデータ
     * @param settings tuneに設定項目
     * @param block tuneに設定されてるblock
     */
    constructor({ api, data, config, block}) {
        this.api = api;
        this.block = block;
        /**
         config:{
            default: "right",
            blocks: {
              header: 'center',
              list: 'right'
            }
          },
         */
        this.settings = config;
        this.data = data || { containerBg: this.getAlignment() }
        this.alignmentSettings = [
            {
				name: 'White BG',
				type: 'header_bg',
				prop: { color: '#fff', tailColor: 'white',},
				title: 'White BG',
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="white"/>
				</svg>				
				`,
			},
			{
				name: 'Primary BG',
				type: 'header_bg',				
				prop: { color: '#33375D', tailColor: 'primary-50',},
				title: 'Primary BG',
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#33375D"/>
				</svg>
				`,
			},
			{
				name: 'Dark BG',
				title: 'Dark BG',
				type: 'header_bg',
				
				prop: { color: '#262626', tailColor: 'gray-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#262626"/>
				</svg>
				`,
			},
			{
				name: 'Transparent BG',
				title: 'Transparent BG',
				type: 'header_bg',
				
				prop: { color: 'transparent', tailColor: 'transparent'},
				icon: `<svg width="15" height="15" viewBox="0 0 15 15" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="5" height="5" fill="#BBF3FF"/>
				<rect x="5" y="5" width="5" height="5" fill="#BBF3FF"/>
				<rect x="10" y="10" width="5" height="5" fill="#BBF3FF"/>
				<rect x="10" width="5" height="5" fill="#BBF3FF"/>
				<rect y="10" width="5" height="5" fill="#BBF3FF"/>
				</svg>
				
					`,
			},


			{
				name: 'Slate - 50 BG',
				title: 'Slate - 50 BG',
				type: 'header_bg',
				prop: { color: '#f8fafc', tailColor: 'slate-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#f8fafc"/>
				</svg>
				`,
			},
			{
				name: 'Slate - 300 BG',
				title: 'Slate - 300 BG',
				type: 'header_bg',
				prop: { color: '#cbd5e1', tailColor: 'slate-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#cbd5e1"/>
				</svg>
				`,
			},{
				name: 'Slate - 600 BG',
				title: 'Slate - 600 BG',
				type: 'header_bg',
				prop: { color: '#475569', tailColor: 'slate-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#475569"/>
				</svg>
				`,
			},
			{
				name: 'Slate - 900 BG',
				title: 'Slate - 900 BG',
				type: 'header_bg',
				prop: { color: '#0f172a', tailColor: 'slate-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#0f172a"/>
				</svg>
				`,
			},




			{
				name: 'Grey - 50 BG',
				title: 'Grey - 50 BG',
				type: 'header_bg',
				prop: { color: '#f9fafb', tailColor: 'gray-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#f9fafb"/>
				</svg>
				`,
			},
			{
				name: 'Grey - 300 BG',
				title: 'Grey - 300 BG',
				type: 'header_bg',
				prop: { color: '#d1d5db', tailColor: 'gray-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#d1d5db"/>
				</svg>
				`,
			},{
				name: 'Grey - 600 BG',
				title: 'Grey - 600 BG',
				type: 'header_bg',
				prop: { color: '#4b5563', tailColor: 'gray-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#4b5563"/>
				</svg>
				`,
			},
			{
				name: 'Grey - 900 BG',
				title: 'Grey - 900 BG',
				type: 'header_bg',
				prop: { color: '#111827', tailColor: 'gray-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#111827"/>
				</svg>
				`,
			},


			{
				name: 'Red - 50 BG',
				title: 'Red - 50 BG',
				type: 'header_bg',
				prop: { color: '#fef2f2', tailColor: 'red-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fef2f2"/>
				</svg>
				`,
			},
			{
				name: 'Red - 300 BG',
				title: 'Red - 300 BG',
				type: 'header_bg',
				prop: { color: '#fca5a5', tailColor: 'red-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fca5a5"/>
				</svg>
				`,
			},{
				name: 'Red - 600 BG',
				title: 'Red - 600 BG',
				type: 'header_bg',
				prop: { color: '#dc2626', tailColor: 'red-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#dc2626"/>
				</svg>
				`,
			},
			{
				name: 'Red - 900 BG',
				title: 'Red - 900 BG',
				type: 'header_bg',
				prop: { color: '#7f1d1d', tailColor: 'red-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#7f1d1d"/>
				</svg>
				`,
			},


			
			{
				name: 'Orange - 50 BG',
				title: 'Orange - 50 BG',
				type: 'header_bg',
				prop: { color: '#fff7ed', tailColor: 'orange-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fff7ed"/>
				</svg>
				`,
			},
			{
				name: 'Orange - 300 BG',
				title: 'Orange - 300 BG',
				type: 'header_bg',
				prop: { color: '#fdba74', tailColor: 'orange-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fdba74"/>
				</svg>
				`,
			},{
				name: 'Orange - 600 BG',
				title: 'Orange - 600 BG',
				type: 'header_bg',
				prop: { color: '#ea580c', tailColor: 'orange-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#ea580c"/>
				</svg>
				`,
			},
			{
				name: 'Orange - 900 BG',
				title: 'Orange - 900 BG',
				type: 'header_bg',
				prop: { color: '#7c2d12', tailColor: 'orange-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#7c2d12"/>
				</svg>
				`,
			},


			{
				name: 'Yellow - 50 BG',
				title: 'Yellow - 50 BG',
				type: 'header_bg',
				prop: { color: '#fefce8', tailColor: 'yellow-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fefce8"/>
				</svg>
				`,
			},
			{
				name: 'Yellow - 300 BG',
				title: 'Yellow - 300 BG',
				type: 'header_bg',
				prop: { color: '#fde047', tailColor: 'yellow-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fde047"/>
				</svg>
				`,
			},{
				name: 'Yellow - 600 BG',
				title: 'Yellow - 600 BG',
				type: 'header_bg',
				prop: { color: '#ca8a04', tailColor: 'yellow-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#ca8a04"/>
				</svg>
				`,
			},
			{
				name: 'Yellow - 900 BG',
				title: 'Yellow - 900 BG',
				type: 'header_bg',
				prop: { color: '#713f12', tailColor: 'yellow-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#713f12"/>
				</svg>
				`,
			},
			
			
			
			{
				name: 'Lime - 50 BG',
				title: 'Lime - 50 BG',
				type: 'header_bg',
				prop: { color: '#f7fee7', tailColor: 'lime-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#f7fee7"/>
				</svg>
				`,
			},
			{
				name: 'Lime - 300 BG',
				title: 'Lime - 300 BG',
				type: 'header_bg',
				prop: { color: '#bef264', tailColor: 'lime-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#bef264"/>
				</svg>
				`,
			},{
				name: 'Lime - 600 BG',
				title: 'Lime - 600 BG',
				type: 'header_bg',
				prop: { color: '#65a30d', tailColor: 'lime-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#65a30d"/>
				</svg>
				`,
			},
			{
				name: 'Lime - 900 BG',
				title: 'Lime - 900 BG',
				type: 'header_bg',
				prop: { color: '#365314', tailColor: 'lime-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#365314"/>
				</svg>
				`,
			},
			
			
			
			{
				name: 'Cyan - 50 BG',
				title: 'Cyan - 50 BG',
				type: 'header_bg',
				prop: { color: '#ecfeff', tailColor: 'cyan-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#ecfeff"/>
				</svg>
				`,
			},
			{
				name: 'Cyan - 300 BG',
				title: 'Cyan - 300 BG',
				type: 'header_bg',
				prop: { color: '#67e8f9', tailColor: 'cyan-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#67e8f9"/>
				</svg>
				`,
			},{
				name: 'Cyan - 600 BG',
				title: 'Cyan - 600 BG',
				type: 'header_bg',
				prop: { color: '#0891b2', tailColor: 'cyan-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#0891b2"/>
				</svg>
				`,
			},
			{
				name: 'Cyan - 900 BG',
				title: 'Cyan - 900 BG',
				type: 'header_bg',
				prop: { color: '#164e63', tailColor: 'cyan-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#164e63"/>
				</svg>
				`,
			},



			{
				name: 'Blue - 50 BG',
				title: 'Blue - 50 BG',
				type: 'header_bg',
				prop: { color: '#eff6ff', tailColor: 'blue-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#eff6ff"/>
				</svg>
				`,
			},
			{
				name: 'Blue - 300 BG',
				title: 'Blue - 300 BG',
				type: 'header_bg',
				prop: { color: '#93c5fd', tailColor: 'blue-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#93c5fd"/>
				</svg>
				`,
			},{
				name: 'Blue - 600 BG',
				title: 'Blue - 600 BG',
				type: 'header_bg',
				prop: { color: '#2563eb', tailColor: 'blue-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#2563eb"/>
				</svg>
				`,
			},
			{
				name: 'Blue - 900 BG',
				title: 'Blue - 900 BG',
				type: 'header_bg',
				prop: { color: '#1e3a8a', tailColor: 'blue-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#1e3a8a"/>
				</svg>
				`,
			},




			{
				name: 'Violet - 50 BG',
				title: 'Violet - 50 BG',
				type: 'header_bg',
				prop: { color: '#f5f3ff', tailColor: 'violet-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#f5f3ff"/>
				</svg>
				`,
			},
			{
				name: 'Violet - 300 BG',
				title: 'Violet - 300 BG',
				type: 'header_bg',
				prop: { color: '#c4b5fd', tailColor: 'violet-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#c4b5fd"/>
				</svg>
				`,
			},{
				name: 'Violet - 600 BG',
				title: 'Violet - 600 BG',
				type: 'header_bg',
				prop: { color: '#7c3aed', tailColor: 'violet-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#7c3aed"/>
				</svg>
				`,
			},
			{
				name: 'Violet - 900 BG',
				title: 'Violet - 900 BG',
				type: 'header_bg',
				prop: { color: '#4c1d95', tailColor: 'violet-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#4c1d95"/>
				</svg>
				`,
			},



			{
				name: 'Pink - 50 BG',
				title: 'Pink - 50 BG',
				type: 'header_bg',
				prop: { color: '#fdf2f8', tailColor: 'pink-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fdf2f8"/>
				</svg>
				`,
			},
			{
				name: 'Pink - 300 BG',
				title: 'Pink - 300 BG',
				type: 'header_bg',
				prop: { color: '#f9a8d4', tailColor: 'pink-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#f9a8d4"/>
				</svg>
				`,
			},{
				name: 'Pink - 600 BG',
				title: 'Pink - 600 BG',
				type: 'header_bg',
				prop: { color: '#db2777', tailColor: 'pink-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#db2777"/>
				</svg>
				`,
			},
			{
				name: 'Pink - 900 BG',
				title: 'Pink - 900 BG',
				type: 'header_bg',
				prop: { color: '#831843', tailColor: 'pink-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#831843"/>
				</svg>
				`,
			},
			
			{
				name: 'Rose - 50 BG',
				title: 'Rose - 50 BG',
				type: 'header_bg',
				prop: { color: '#fff1f2', tailColor: 'rose-50'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fff1f2"/>
				</svg>
				`,
			},
			{
				name: 'Rose - 300 BG',
				title: 'Rose - 300 BG',
				type: 'header_bg',
				prop: { color: '#fda4af', tailColor: 'rose-300'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#fda4af"/>
				</svg>
				`,
			},{
				name: 'Rose - 600 BG',
				title: 'Rose - 600 BG',
				type: 'header_bg',
				prop: { color: '#e11d48', tailColor: 'rose-600'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#e11d48"/>
				</svg>
				`,
			},
			{
				name: 'Rose - 900 BG',
				title: 'Rose - 900 BG',
				type: 'header_bg',
				prop: { color: '#881337', tailColor: 'rose-900'},
				icon: `<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
				<rect width="20" height="20" rx="5" fill="#881337"/>
				</svg>
				`,
			},
        ];

    }

    /**
     * block自体をwrapしてくれる
     * constructorで与えられたalignmentを代入しようとすると、holderが確定してなく
     * renderでやろうとすると、tuneを表示時に処理が走る
     * @param blockContent
     */
    wrap(blockContent) {
        this.wrapper = this.make("div");
        this.wrapper.classList.toggle(`bg-${this.data.containerBg}`)
        this.wrapper.append(blockContent)
        return this.wrapper
    }

    /**
     * rendering block tune
     * @returns {*}
     */
    render() {
        const wrapper = this.make("div");
        // const wrapperColorsText = this.make('div');
		wrapper.classList.add('cdx-settings-colorsWrapper');

        // this.alignmentSettings.forEach((tune, index, elements) => {
		// 	let button = document.createElement('div');
			
		// 	button.classList.add('cdx-settings-button-containerBg');
		// 	button.setAttribute('title', tune.title);
		// 	button.innerHTML = tune.icon;
		
        //     wrapperColorsText.appendChild(button);
		
		

		// 	button.addEventListener('click', () => {
		// 		// this._toggleTune(tune.prop);

        //         // for (var item of document.querySelectorAll('.cdx-settings-button-containerBg--active')) {
        //         //     item.classList.remove('cdx-settings-button-containerBg--active');
        //         // }
        //         // button.classList.toggle('cdx-settings-button-containerBg--active');
        //         this.data = {
        //             containerBg: tune.prop.tailColor
        //         }
        //         elements.forEach((el, i) => {
        //             const {tailColor} = this.alignmentSettings[i].prop;
        //             el.classList.toggle(this.api.styles.settingsButtonActive, tailColor === this.data.containerBg);
        //             //toggle alignment style class for block
        //             this.wrapper.classList.toggle(`bg-${tailColor}`, name === this.data.containerBg)
        //         });
		// 	  });
		// });
		
	
		

        this.alignmentSettings.map(align => {
            const button = document.createElement('button');
            button.classList.add(this.api.styles.settingsButton);
            button.innerHTML = align.icon;
            button.type = 'button';

            button.classList.toggle(this.api.styles.settingsButtonActive, align.prop.tailColor === this.data.containerBg);
            wrapper.appendChild(button);
            return button
        }).forEach((element, index, elements) => {
            element.addEventListener('click', () => {
                this.data = {
                    containerBg: this.alignmentSettings[index].prop.tailColor
                }
                elements.forEach((el, i) => {
                    const {tailColor} = this.alignmentSettings[i].prop;
                    el.classList.toggle(this.api.styles.settingsButtonActive, tailColor === this.data.containerBg);
                    //toggle alignment style class for block
                    this.wrapper.classList.toggle(`bg-${tailColor}`, tailColor === this.data.containerBg)
                    // this.wrapper.classList.toggle(this._CSS.alignment[name], name === this.data.alignment)
                });
            });
        });
        // wrapper.appendChild(wrapperColorsText);
        return wrapper;
    }
    /**
     * save
     * @returns {*}
     */
    save() {
        return this.data;
    }

    make(tagName, classNames = null, attributes = {}) {
        const el = document.createElement(tagName);
      
        if (Array.isArray(classNames)) {
          el.classList.add(...classNames);
        } else if (classNames) {
          el.classList.add(classNames);
        }
      
        for (const attrName in attributes) {
          el[attrName] = attributes[attrName];
        }
        return el;
    }
}

module.exports = containerBgTune;
