// @silver0542/directus-extension-cma-components-editor/tailwind.config.js
module.exports = {
  content: [
    "./index.html",
    "./src/**/*.{vue,js,ts,jsx,tsx}",
    "./node_modules/flowbite/**/*.js",
    "lg:flex-1",
      "lg:grow",
      "lg:basis-0",
      "lg:min-w-0",
      {
        pattern: /(bg|text|border)-./,
      },
      {
        pattern: /(w|items|p|pt|pb|pl|pr|py|px|m|mt|mb|ml|mr|my|mx)-./,
        variants: ["xs", "sm", "md", "lg", "xl"],
      },
  ],
  theme: {
    extend: {
      colors: {
        primary: {
          50: "#33375d",
          100: "#eee",
          200: "#eaebf4",
          300: "#d5d7e8",
          400: "rgba(51,55,93, .7)",
          500: "#222222",
          600: "#2563eb",
          700: "#1d4ed8",
          800: "#1e40af",
          900: "#1e3a8a",
        },
        blue: {
          // 50: "#33375d",
          // 50: "#000",
          // 100: "#50B8FC",
          // 200: "#bfdbfe",
          300: "#93c5fd",
          //400: "#60a5fa",
          500: "#3b82f6",
          600: "#2563eb",
          700: "#1d4ed8",
          800: "#1e40af",
          // 900: "#1e3a8a",
        },
      },
    },
  },
  plugins: [
    require('flowbite/plugin')
  ],
  
}
